src/datasources/balances-api/valk-balances-api.service.ts

```diff
@@ -0,0 +1,143 @@
+    this.defaultNotFoundExpirationTimeSeconds =
+      this.configurationService.getOrThrow<number>(
+        'expirationTimeInSeconds.notFound.default',
+      );
+    this.chainsConfiguration = this.configurationService.getOrThrow<
+      Record<number, ChainAttributes>
+    >('balances.providers.valk.chains');
+  }
+
+  async getBalances(args: {
+    chainId: string;
+    safeAddress: string;
+    fiatCode: string;
+  }): Promise<Balance[]> {
+    try {
+      const cacheDir = CacheRouter.getValkBalancesCacheDir(args);
+      const chainName = this.getChainName(args.chainId);
+      const url = `${this.baseUri}/balances/token/${args.safeAddress}?chain=${chainName}`;
+      const valkBalances = await this.dataSource.get<ValkBalance[]>({
+        cacheDir,
+        url,
+        notFoundExpireTimeSeconds: this.defaultNotFoundExpirationTimeSeconds,
+        networkRequest: { headers: { Authorization: `${this.apiKey}` } },
+        expireTimeSeconds: this.defaultExpirationTimeInSeconds,
+      });
+      return this.mapBalances(valkBalances, args.fiatCode);
+    } catch (error) {
+      throw new DataSourceError(
+        `Error getting ${args.safeAddress} balances from provider: ${asError(error).message}}`,
+      );
+    }
+  }
+
+  mapBalances(valkBalances: ValkBalance[], fiatCode: string): Balance[] {
+    return valkBalances.map((valkBalance) => {
+      const price = valkBalance.prices[fiatCode.toUpperCase()] ?? null;
+      const balance = getNumberString(valkBalance.balance);
+      const fiatBalance = getNumberString(
+        (valkBalance.balance / Math.pow(10, valkBalance.decimals)) * price,
+      );
+      const fiatConversion = getNumberString(price);
+
+      // Valk returns a string representing the native coin (e.g.: 'eth') as
+      // token_address for native coins balances. An hex is returned for ERC20 tokens.
+      if (isHex(valkBalance.token_address)) {
+        return {
+          tokenAddress: valkBalance.token_address,
+          token: {
+            name: valkBalance.name,
+            symbol: valkBalance.symbol,
+            decimals: valkBalance.decimals,
+            logoUri: valkBalance.logo ?? '',
+          },
+          balance,
+          fiatBalance,
+          fiatConversion,
+        } as Erc20Balance & {
+          fiatBalance: string | null;
+          fiatConversion: string | null;
+        };
+      } else {
+        return {
+          tokenAddress: null,
+          token: null,
+          balance,
+          fiatBalance,
+          fiatConversion,
+        } as NativeBalance & {
+          fiatBalance: string | null;
+          fiatConversion: string | null;
+        };
+      }
+    });
+  }
+
+  getChainName(chainId: string): string {
+    const chainName = this.chainsConfiguration[Number(chainId)]?.chainName;
+    if (!chainName)
+      throw Error(
+        `Chain ${chainId} balances retrieval via Valk is not configured`,
+      );
+    return chainName;
+  }
+
+  async clearBalances(args: {
```

Comment:
Nit: I'd try to be declarative here with the namingg as well, e.g. `clearExternalBalancesApi`.

Reply:
Hmm, I guess that would also imply to rename `getBalances` to `getExternalBalances`. No strong opinion, but from a domain-side point of view (as this interface serves as API for the domain), I'd say we should abstract the precise origin of the balances (external provider vs. Transaction Service). And also, eventually, all the balances could become "external" so no distinction should be necessary. Thoughts @iamacook @fmrsabino? 🙂

Reply:
Fair point. As this is in a provider-specific service then let's it as is.